### PR TITLE
feat(memory): add delete_memory tool to MemoryManager (#3577)

### LINKED ISSUE #3577: [FEATURE] Add `delete_memory` / `forget_memory` tool to MemoryManager
### SDK Language

Both / Not language-specific

### Problem Statement

The `MemoryManager` currently provides `search_memory` and `add_memory` tools for agents, but there is no way for an agent to delete or forget previously stored memories. Once a memory is written to a writable store (e.g., `BedrockKnowledgeBaseStore`), it persists indefinitely with no SDK-level mechanism to remove it.

This creates problems when:
- A user asks the agent to forget something they previously shared
- Memories become stale or incorrect and need to be removed
- Users want control over what personal data the agent retains (privacy/GDPR considerations)
- Extracted memories contain errors that the agent should be able to correct by deletion

### Proposed Solution

Add a `delete_memory` tool (or `forget_memory`) to the `MemoryManager` that:

1. Accepts a `memory_id` or `query` parameter to identify which memory to delete
2. Operates on writable stores only (similar to `add_memory`)
3. Delegates deletion to the store's backend (e.g., removing the S3 object and triggering a KB re-sync for `BedrockKnowledgeBaseStore`)

Example interface:

```python
# Agent-invocable tool
def delete_memory(memory_id: str, store: str | None = None) -> str:
    """Delete a specific memory from a writable store."""
    ...
```

The store interface would need a `delete` method:

```python
class BedrockKnowledgeBaseStore:
    def delete(self, memory_id: str) -> bool:
        # Remove from S3 data source
        # Trigger ingestion job to re-sync
        ...
```

A `MemoryDeleteToolConfig` (similar to `MemoryAddToolConfig`) could control whether the tool is exposed to the agent.

### Use Case

**Use Case:**

1. **User-requested forgetting:** A user tells the agent "Forget that my favorite color is blue" — the agent should be able to remove that specific memory from the user's store.

2. **Privacy compliance:** Users request deletion of personal data. The agent can act on this immediately rather than requiring manual admin intervention.

3. **Memory correction:** Rather than adding a contradicting memory ("Actually my favorite color is NOT blue"), the agent can cleanly delete the incorrect one and optionally store the correction.

4. **Session cleanup:** An agent that stores temporary context as memories during a long task can clean up after itself.

### Alternatives Solutions

**Alternative Solutions:**

1. **Overwrite with contradicting memory** — The agent writes a new memory that supersedes the old one. Pros: works today. Cons: accumulates stale data, relies on retrieval ranking to surface the latest, no actual deletion.

2. **Custom `@tool` wrapping S3/KB APIs** — Build a custom delete tool outside the SDK that directly deletes from the backing store and triggers re-sync. Pros: works today. Cons: tightly couples to the store implementation, duplicates logic the SDK should own, different for every store backend.

3. **Admin API / out-of-band deletion** — Expose an HTTP endpoint for users to manage their memories outside the agent. Pros: gives users direct control. Cons: not agent-driven, requires separate UI/UX work, doesn't solve the in-conversation "forget this" use case.

### Additional Context

**Additional Context:**

The `search_memory` tool already returns results that could include a `memory_id` or similar identifier. This identifier could be passed to `delete_memory` to target a specific entry.

For `BedrockKnowledgeBaseStore` backed by S3, deletion would involve:
1. Identifying the S3 object (by memory ID / key)
2. Calling `s3:DeleteObject`
3. Triggering `bedrock-agent:StartIngestionJob` to re-sync the KB

The `add_memory` flow already handles writing to S3 and syncing, so the inverse operation should follow the same pattern.

### PR DESCRIPTION
**Pull request by lifelmy**
## Description

Adds a `delete_memory` / forget capability to the `MemoryManager`, addressing #3577. Agents (and callers) can now remove a specific previously-stored memory instead of only being able to add or overwrite one.

### What changed

- **`MemoryStore` Protocol** gains an optional `async def delete(self, memory_id: str)`. Implementing it opts a writable store into deletion. `MemoryEntry` now carries an opaque `memory_id` that the originating store sets on search results, so a search result can later be passed back to `delete`.
- **`MemoryManager`**:
  - New `delete_tool_config` constructor param (opt-in, default `False`; `True` = all writable delete-capable stores; `MemoryDeleteToolConfig` to restrict/rename).
  - New programmatic `async def delete(memory_id, options)` that targets delete-capable writable stores, runs deletes concurrently, and surfaces per-store failures as an `AggregateMemoryError` — mirroring the existing `add` flow.
  - New agent-invocable `delete_memory(memory_id, stores=None)` tool returning `{"deleted": memory_id}`, reusing the same store-scoping/allowlist/warning behavior as `add_memory`.
  - `search_memory` results now include `memory_id` so the model can target a delete.
- **`FileMemoryStore`** sets `memory_id` (its storage key) on search results and implements `delete` (no-op on a missing key, matching the storage backend).
- **Tracer** gains `memory.delete` spans, consistent with `memory.add` / `memory.search`.

Deletion is opt-in and never bypasses read-only stores. The `memory_id` is opaque to the manager; `BedrockKnowledgeBaseStore` and other stores can implement `delete` in follow-ups per the issue's S3/re-sync notes.

## Related Issues

Closes #3577

## Testing

Run from `strands-py/`:

- `hatch test tests/strands/memory/test_memory_manager.py tests/strands/vended_memory_stores/file_memory_store/test_store.py` — 168 passed (adds constructor-validation, programmatic `delete`, `delete_memory` tool scoping/return-value, and FileMemoryStore delete tests).
- `hatch test tests/strands/telemetry/` — 172 passed (covers the new delete spans path).
- `ruff check` — All checks passed. `ruff format` applied to the changed files.

New public API is additive and backwards-compatible (delete tool defaults off).

## Note on authorship

I use a coding assistant to help implement, and I review and take responsibility for the final change.

## Checklist

- [x] I have read the CONTRIBUTING document
- [x] Related issue linked
- [x] Tests added/updated and passing locally
- [x] Backwards compatible (opt-in, additive)

### CODE REVIEW

File: strands-py/src/strands/memory/memory_manager.py [hunk 1]
```diff
@@ -569,6 +710,45 @@ async def add_memory(entries: list[str], stores: list[str] | None = None) -> dic
...
+                    available stores.
+
+            Returns:
+                A summary of the delete (``{"deleted": memory_id}``).
+            """
+            # @tool validation does not enforce non-empty strings, so guard here.
+            if not memory_id:
+                raise ValueError("MemoryManager: delete_memory requires a memory_id")
+
+            targets = self._resolve_tool_targets(scoped_names, stores)
```
Comment (arielnabavian) [inline strands-py/src/strands/memory/memory_manager.py]: [arielnab-ai]
**concern** — When the model omits `stores`, the tool sends the same `memory_id` to every store in scope. But `memory_id` is store-native, not global. Two stores can hold different entries under the same id.

Concrete case, with the two-store setup from the docs (`personal` and `team`, both `FileMemoryStore`): each store names its file after the first heading, so both hold `user-preferences.md`. `_NamespacedStorage.search` strips the namespace prefix, so both entries report `memory_id="user-preferences.md"`. The user asks to forget the preferences in `personal`. The model calls `delete_memory(memory_id="user-preferences.md")`. The tool deletes the `team` file too. The description on line 721 tells the model that this is the normal path.

Suggested fix: when more than one store is in scope and `stores` is omitted, raise a `ValueError` that tells the model to pass the `store_name` from the search result. Or take one required `store` argument, as the issue proposed. The programmatic `MemoryManager.delete` has the same default, but a developer controls that call.

File: strands-py/src/strands/memory/types.py [hunk 2]
```diff
@@ -296,6 +330,18 @@ async def add_messages(self, messages: list[Message], context: AddMessagesContex
         """
         ...
 
+    async def delete(self, memory_id: str) -> Any:
```
Comment (arielnabavian) [inline strands-py/src/strands/memory/types.py]: [arielnab-ai]
**concern** — `delete` returns `Any`, and the manager ignores the value. So the tool cannot tell "removed" from "nothing matched". `Storage.delete` is a no-op for a missing key, so `FileMemoryStore.delete` never reports a miss. The tool then returns `{"deleted": memory_id}` (memory_manager.py line 744) for a wrong or stale id, and the model tells the user that the memory is gone.

This is also hard to change later. Once third-party stores implement `delete -> Any`, the manager cannot start to read the return value without a breaking change.

Suggested fix: make the contract `async def delete(self, memory_id: str) -> bool`, `True` when an entry was removed. The issue proposed this shape. Then `MemoryManager.delete` can return the names of the stores that removed an entry. The tool can return `{"deleted": memory_id, "stores": [...]}`, or raise a "not found" error when no store removed anything.

File: strands-py/src/strands/memory/memory_manager.py [hunk 3]
```diff
@@ -498,6 +637,8 @@ async def search_memory(
                 item: dict[str, Any] = {"content": entry.content}
                 if entry.store_name:
                     item["store_name"] = entry.store_name
+                if entry.memory_id is not None:
+                    item["memory_id"] = entry.memory_id
```
Comment (arielnabavian) [inline strands-py/src/strands/memory/memory_manager.py]: [arielnab-ai]
**concern** — No test covers `memory_id` on the search path. The new tests set `memory_id` only on delete calls. No fake store returns a `MemoryEntry` with a `memory_id`, so nothing asserts that `MemoryManager.search` (line 427) keeps the id, or that this payload includes it. That path is the only way the model gets an id to delete. Add one test with a store entry that has a `memory_id` and one that has `None`, and assert the tool payload for each.

File: strands-py/src/strands/telemetry/tracer.py [hunk 4]
```diff
@@ -1126,6 +1126,56 @@ def end_memory_add_span(
         attributes: dict[str, AttributeValue] = {"memory.store.failure_count": store_failure_count}
         self._end_span(span, attributes, error)
 
+    def start_memory_delete_span(
```
Comment (arielnabavian) [inline strands-py/src/strands/telemetry/tracer.py]: [arielnab-ai]
**concern** — `test_tracer.py` has tests for the add and search spans (`test_start_memory_add_span`, `test_end_memory_add_span_with_error`, and so on) but none for `start_memory_delete_span` or `end_memory_delete_span`. The memory manager tests run this code but assert nothing about the span. Add the matching tests: the `memory.id`, `memory.store.names`, and `memory.store.count` attributes on start, and `memory.store.failure_count` plus the error status on end.

File: strands-py/src/strands/vended_memory_stores/file_memory_store/store.py [hunk 5]
```diff
@@ -200,3 +201,16 @@ async def add(self, content: str, metadata: Metadata | None = None) -> str:
             await self._storage.write(key, merged.encode("utf-8"))
 
         return key
+
+    async def delete(self, memory_id: str) -> None:
```
Comment (arielnabavian) [inline strands-py/src/strands/vended_memory_stores/file_memory_store/store.py]: [arielnab-ai]
**consider** — In this store, the delete unit is a topic file, not one fact. `add` merges every fact with the same heading into one file (lines 192-201), and the default extraction prompt asks for headings like "# User preferences". So "forget that my favorite color is blue" deletes all user preferences, because the only `memory_id` the model can get is `user-preferences.md`. That is the first use case in the issue.

Fact-level deletion can be a follow-up. For this PR, say this in the tool description and in this docstring, so the model and the developer know that the whole entry shown by `search_memory` goes away.
